File: js/sheets/weeks.js
// Season-boundary + week-list logic: auto-computes which Shabbosim belong on a
// Kayitz (summer) or Choref (winter) sheet for a given Hebrew year, mirroring the
// workbook's own P5 formula (SEQUENCE + FILTER on HAS_PARSHA<>"") but without
// requiring a manually-entered start date or week count.
import { dateFromHebrew, hasParsha, hasSpecialParsha, hasYomTov, isYomTovOrCholHamoed, hebrewDateExtended, excelWeekday } from '../hebrew-calendar.js';
import { dateFromSerial, excelSerial } from '../zmanim/solar.js';
import { inSpringDstWindow } from './common.js';

const MAX_WEEKS = 60; // safety cap, well above any real season's length

/** Season(hebrewYear) date-range boundaries — Kayitz(Y): Pesach(Y) -> Sukkos(Y+1);
 *  Choref(Y): Sukkos(Y) -> Pesach(Y), both within AM year Y. Factored out of
 *  computeSeasonWeeks so the "which year is next" helpers below can use the same
 *  boundaries without needing a parsha table. */
function seasonStartSerial(season, hebrewYear) {
  return season === 'kayitz' ? dateFromHebrew(15, 1, hebrewYear) : dateFromHebrew(15, 7, hebrewYear);
}
function seasonEndSerial(season, hebrewYear) {
  return season === 'kayitz' ? dateFromHebrew(15, 7, hebrewYear + 1) : dateFromHebrew(15, 1, hebrewYear);
}

/**
 * @param {'kayitz'|'choref'} season
 * @param {number} hebrewYear AM year anchoring the season (see README for the exact
 *   convention: Kayitz(Y) runs Pesach(Y) -> Sukkos(Y+1); Choref(Y) runs Sukkos(Y) -> Pesach(Y), both within AM year Y)
 * @param {object} settings
 * @param {object} tables {parshaChutz, parshaEY, parshaNames}
 * @returns {{startSerial:number, endSerial:number, weeks: Array<{serial:number,date:Date,parsha:string,specialParsha:string}>}}
 */
export function computeSeasonWeeks(season, hebrewYear, settings, tables) {
  const startSerial = seasonStartSerial(season, hebrewYear);
  const endSerial = seasonEndSerial(season, hebrewYear);

  let d = Math.ceil(startSerial);
  while (excelWeekday(d) !== 7) d++;

  const weeks = [];
  let guard = 0;
  while (d <= endSerial && guard < MAX_WEEKS) {
    const parsha = hasParsha(d, settings, tables);
    if (parsha) {
      weeks.push({ serial: d, date: dateFromSerial(d), parsha, specialParsha: hasSpecialParsha(d, settings) });
    }
    d += 7;
    guard++;
  }
  return { startSerial, endSerial, weeks };
}

function isCholHamoedAnchor(day, settings, specialDaysTable) {
  return /Chol Hamoed|חול המועד|Hoshana Rabbah|הושענה רבה/.test(hasYomTov(day, settings, specialDaysTable));
}

function anyRegularDay(fromSerial, toSerial, settings, specialDaysTable) {
  for (let day = fromSerial; day <= toSerial; day++) {
    if (!isYomTovOrCholHamoed(day, settings, specialDaysTable)) return true;
  }
  return false;
}

/** "Chol Hamoed Pesach" / "חול המועד סוכות" / "Hoshana Rabbah" -> the plain holiday
 *  name ("Pesach" / "סוכות" / "סוכות") — see the Chol Hamoed row below, which reuses
 *  whatever this Shabbos's own hasYomTov() name is but without the "Chol Hamoed"/
 *  "Hoshana Rabbah" framing, since the row is standing in for the holiday as a whole. */
function holidayNameFor(day, settings, specialDaysTable) {
  const name = hasYomTov(day, settings, specialDaysTable);
  if (/^(Hoshana Rabbah|הושענה רבה)$/.test(name)) return /[֐-׿]/.test(name) ? 'סוכות' : 'Succos';
  return name.replace(/^(Chol Hamoed |חול המועד )/, '');
}

/** Week list for a Weekday chart covering the same season date range as
 *  computeSeasonWeeks, but with a different inclusion rule: a week is included as long
 *  as at least one of its Sun-Fri days (the days a Weekday chart actually schedules) is
 *  a normal day — not Yom Tov, not Chol Hamoed. That's a superset of the Shabbos
 *  chart's own week list: a week whose Shabbos falls on Yom Tov (and so has no parsha,
 *  excluded from computeSeasonWeeks) can still need a Weekday-chart row if, say, only
 *  Thursday and Friday of that week are Yom Tov and the rest are regular days.
 *  Each week is still anchored to its Shabbos `serial` (Saturday) for consistency with
 *  the Shabbos weeks list; `parsha` falls back to that Shabbos's own Yom Tov name (e.g.
 *  "ראש השנה") when there's no regular parsha to label the row with.
 *
 *  Season boundaries (Pesach/Sukkos) essentially never line up with the fixed 7-day
 *  Saturday spacing this loop walks in, which leaves a "leftover" stretch of up to 6
 *  regular days between the last Saturday-anchored week and the boundary itself (e.g.
 *  the days between שבת הגדול and ליל פסח) — see the trailing-gap check after the main
 *  loop, which folds that stretch onto *this* (the outgoing/earlier) season as one more
 *  row, per "a week that falls between two charts belongs on the earlier one".
 *
 *  A Saturday that lands ON Chol Hamoed/Hoshana Rabbah (e.g. Shabbos Chol Hamoed Pesach)
 *  is excluded from becoming its own row in the *incoming* season's own loop above (its
 *  backward window would otherwise mix genuine pre-Yom-Tov regular days, already
 *  claimed by the outgoing chart's trailing row, together with actual Chol Hamoed days)
 *  — but it isn't dropped: the trailing-Chol-Hamoed check right after also folds it onto
 *  the *outgoing* chart as one more row, labeled with the holiday's plain name (Pesach's
 *  own Chol Hamoed row lands on the חורף chart; Sukkos's on the קיץ chart). */
export function computeWeekdayWeeks(season, hebrewYear, settings, tables) {
  const startSerial = seasonStartSerial(season, hebrewYear);
  const endSerial = seasonEndSerial(season, hebrewYear);

  let d = Math.ceil(startSerial);
  while (excelWeekday(d) !== 7) d++;

  const weeks = [];
  let guard = 0;
  while (d <= endSerial && guard < MAX_WEEKS) {
    // d === startSerial: the season's own start boundary landed exactly on Shabbos
    // (e.g. Sukkos falling on a Saturday) — its backward-attached weekdays are still
    // within the *outgoing* season's territory, already covered by its own trailing-gap
    // row below. Without this, both seasons would independently print an identical row.
    const isOwnStartBoundary = d === startSerial;
    if (!isOwnStartBoundary && !isCholHamoedAnchor(d, settings, tables.specialDays) && anyRegularDay(d - 6, d - 1, settings, tables.specialDays)) {
      const parsha = hasParsha(d, settings, tables) || hasYomTov(d, settings, tables.specialDays);
      weeks.push({ serial: d, date: dateFromSerial(d), parsha, specialParsha: hasSpecialParsha(d, settings) });
    }
    d += 7;
    guard++;
  }

  // Trailing gap: the regular days (if any) between the last Saturday-anchored week
  // above and the season's own end boundary — see the function comment. Anchored at
  // endSerial itself (not a real Shabbos, just a stand-in date/key for this row) and
  // labeled with whatever Yom Tov starts there, same fallback as any other
  // Yom-Tov-only row above. When the boundary itself was exactly Shabbos, the main loop
  // already picked it up directly and this gap comes out empty — no double-counting.
  const gapStart = d - 7 + 1;
  const gapEnd = endSerial - 1;
  if (gapStart <= gapEnd && anyRegularDay(gapStart, gapEnd, settings, tables.specialDays)) {
    const parsha = hasParsha(endSerial, settings, tables) || hasYomTov(endSerial, settings, tables.specialDays);
    weeks.push({ serial: endSerial, date: dateFromSerial(endSerial), parsha, specialParsha: hasSpecialParsha(endSerial, settings) });
  }

  // `d` is now the first Saturday *after* the boundary (Saturdays fall on the same
  // fixed 7-day cadence no matter which season's math found them, so this is exactly
  // the same date the incoming season's own loop would land on as its own first
  // candidate) — see the function comment for why a Chol-Hamoed/Hoshana-Rabbah Shabbos
  // there becomes a row here instead of there. It's folded into the trailing-gap row
  // above rather than added separately whenever that row already carries the same
  // holiday's name: the two stretches are the run-up to, and the middle of, one single
  // Yom Tov, so the chart should carry one row for it, not a pair of identical ones.
  const cholHamoedName = isCholHamoedAnchor(d, settings, tables.specialDays) ? holidayNameFor(d, settings, tables.specialDays) : '';
  if (cholHamoedName && weeks[weeks.length - 1]?.parsha !== cholHamoedName) {
    weeks.push({ serial: d, date: dateFromSerial(d), parsha: cholHamoedName, specialParsha: '' });
  }

  return { startSerial, endSerial, weeks };
}

/** Where a שבת חורף season's weeks cross the *spring* DST cutover (2nd Sunday of
 *  March — not the fall one near Sukkos at the season's start). From that week on,
 *  the season needs an actual שבת קיץ chart (not just a couple of extra columns) —
 *  the shul still davens on a "summer" schedule through Pesach once the clock springs
 *  forward. Returns the index in `weeks` of the first week on/after the cutover
 *  (weeks.length if the whole season is still before it — shouldn't happen in
 *  practice, since Pesach always falls after the 2nd Sunday of March). */
export function splitChorefAtSpringCutover(weeks, settings) {
  const idx = weeks.findIndex((w) => inSpringDstWindow(w.date, settings));
  return idx === -1 ? weeks.length : idx;
}

/** The smallest hebrewYear for `season` whose date range hasn't already fully elapsed
 *  (its end is still today or later) — i.e. the soonest occurrence of that season still
 *  worth preparing a schedule for. Used to keep the Generate form's year field from
 *  ever defaulting to an already-passed season. */
export function nextAvailableYearFor(season, settings) {
  const today = excelSerial(new Date());
  let y = hebrewDateExtended(today, settings.useGregorianBefore1582).year - 1; // step back one to not overshoot a season that started in a lower-numbered year
  for (let i = 0; i < 6 && seasonEndSerial(season, y) < today; i++) y++;
  return y;
}

/** Which season+year the Generate form should default to: the *next* season
 *  chronologically after whichever one contains today — a schedule is always being
 *  prepared ahead of time for the upcoming season, not the one currently in progress.
 *  E.g. if today falls within a קיץ season, default to the חורף season right after it
 *  (never the קיץ season itself, and never a season that's already over). */
export function defaultSeasonAndYear(settings) {
  const today = excelSerial(new Date());
  const y0 = hebrewDateExtended(today, settings.useGregorianBefore1582).year;
  const sukkosY0 = dateFromHebrew(15, 7, y0);
  const pesachY0 = dateFromHebrew(15, 1, y0);
  const sukkosY0plus1 = dateFromHebrew(15, 7, y0 + 1);

  let currentSeason, currentYear;
  if (today < sukkosY0) {
    currentSeason = 'kayitz';
    currentYear = y0 - 1; // still in last cycle's קיץ — Sukkos(y0) hasn't happened yet
  } else if (today < pesachY0) {
    currentSeason = 'choref';
    currentYear = y0;
  } else if (today < sukkosY0plus1) {
    currentSeason = 'kayitz';
    currentYear = y0;
  } else {
    currentSeason = 'choref';
    currentYear = y0 + 1;
  }

  return currentSeason === 'choref' ? { season: 'kayitz', hebrewYear: currentYear } : { season: 'choref', hebrewYear: currentYear + 1 };
}

File: desktop/zmanimboard/engine/sheets/weeks.py
"""Season boundaries and week lists, ported from js/sheets/weeks.js.

Which Shabbosim belong on a קיץ or חורף chart for a given Hebrew year, mirroring the
workbook's own P5 formula but without needing a manually entered start date or week count.
"""
import math
import re
from dataclasses import dataclass
from datetime import date

from ..hebrew_calendar import (
    date_from_hebrew,
    excel_weekday,
    has_parsha,
    has_special_parsha,
    has_yom_tov,
    hebrew_date_extended,
    is_yom_tov_or_chol_hamoed,
)
from ..solar import date_from_serial, excel_serial
from .common import in_spring_dst_window

MAX_WEEKS = 60  # a safety cap, well above any real season's length


@dataclass(frozen=True)
class Week:
    serial: int
    date: date
    parsha: str
    special_parsha: str


@dataclass(frozen=True)
class SeasonWeeks:
    start_serial: int
    end_serial: int
    weeks: list


def season_start_serial(season: str, hebrew_year: int) -> int:
    """קיץ(Y) opens at Pesach(Y); חורף(Y) opens at Sukkos(Y)."""
    return date_from_hebrew(15, 1, hebrew_year) if season == "kayitz" else date_from_hebrew(15, 7, hebrew_year)


def season_end_serial(season: str, hebrew_year: int) -> int:
    """קיץ(Y) closes at Sukkos(Y+1); חורף(Y) closes at Pesach(Y), both inside AM year Y."""
    return date_from_hebrew(15, 7, hebrew_year + 1) if season == "kayitz" else date_from_hebrew(15, 1, hebrew_year)


def _first_saturday_on_or_after(serial: float) -> int:
    d = math.ceil(serial)
    while excel_weekday(d) != 7:
        d += 1
    return d


def compute_season_weeks(season: str, hebrew_year: int, s, tables) -> SeasonWeeks:
    start_serial = season_start_serial(season, hebrew_year)
    end_serial = season_end_serial(season, hebrew_year)

    d = _first_saturday_on_or_after(start_serial)
    weeks, guard = [], 0
    while d <= end_serial and guard < MAX_WEEKS:
        parsha = has_parsha(d, s, tables)
        if parsha:
            weeks.append(Week(d, date_from_serial(d), parsha, has_special_parsha(d, s)))
        d += 7
        guard += 1
    return SeasonWeeks(start_serial, end_serial, weeks)


_CHOL_HAMOED_ANCHOR = re.compile(r"Chol Hamoed|חול המועד|Hoshana Rabbah|הושענה רבה")


def _is_chol_hamoed_anchor(day: int, s, special_days_table) -> bool:
    return bool(_CHOL_HAMOED_ANCHOR.search(has_yom_tov(day, s, special_days_table)))


def _any_regular_day(from_serial: int, to_serial: int, s, special_days_table) -> bool:
    return any(not is_yom_tov_or_chol_hamoed(day, s, special_days_table) for day in range(from_serial, to_serial + 1))


def _holiday_name_for(day: int, s, special_days_table) -> str:
    """"Chol Hamoed Pesach" or "הושענה רבה" to the plain holiday name, for the Chol Hamoed
    row below: it stands in for the holiday as a whole rather than for one of its days."""
    name = has_yom_tov(day, s, special_days_table)
    if re.fullmatch(r"Hoshana Rabbah|הושענה רבה", name):
        return "סוכות" if re.search(r"[֐-׿]", name) else "Succos"
    return re.sub(r"^(Chol Hamoed |חול המועד )", "", name)


def compute_weekday_weeks(season: str, hebrew_year: int, s, tables) -> SeasonWeeks:
    """The Weekday chart's own week list, over the same season range but with a different
    inclusion rule: a week counts as long as at least one of its Sunday to Friday days, the
    days a Weekday chart actually schedules, is a normal day.

    That is a superset of the Shabbos chart's list. A week whose Shabbos falls on Yom Tov
    has no parsha and is excluded there, but can still need a Weekday row if only its
    Thursday and Friday are Yom Tov and the rest are ordinary days. Each week is still
    anchored to its Shabbos, and the label falls back to that Shabbos's Yom Tov name when
    there is no parsha to name the row with.

    Season boundaries essentially never line up with the fixed seven day Saturday spacing,
    which leaves a stretch of up to six regular days between the last Saturday anchored
    week and the boundary itself, the days between שבת הגדול and ליל פסח for instance. The
    trailing gap check folds that onto the outgoing season as one more row, per "a week
    falling between two charts belongs to the earlier one".

    A Saturday landing on Chol Hamoed is excluded from the incoming season's own loop, its
    backward window otherwise mixing genuine pre-Yom-Tov days already claimed by the
    outgoing chart with actual Chol Hamoed days. It is not dropped: the check after the
    loop folds it onto the outgoing chart, named for the holiday.
    """
    start_serial = season_start_serial(season, hebrew_year)
    end_serial = season_end_serial(season, hebrew_year)

    d = _first_saturday_on_or_after(start_serial)
    weeks, guard = [], 0
    while d <= end_serial and guard < MAX_WEEKS:
        # d == start_serial: the season's own start boundary landed exactly on Shabbos, so
        # its backward attached weekdays are still the outgoing season's territory, already
        # covered by its trailing gap row. Without this both seasons print the same row.
        is_own_start_boundary = d == start_serial
        if (
            not is_own_start_boundary
            and not _is_chol_hamoed_anchor(d, s, tables.special_days)
            and _any_regular_day(d - 6, d - 1, s, tables.special_days)
        ):
            parsha = has_parsha(d, s, tables) or has_yom_tov(d, s, tables.special_days)
            weeks.append(Week(d, date_from_serial(d), parsha, has_special_parsha(d, s)))
        d += 7
        guard += 1

    # The trailing gap: the regular days, if any, between the last Saturday anchored week
    # and the season's end boundary. Anchored at end_serial itself, which is a stand-in
    # date rather than a real Shabbos, and named for whatever Yom Tov starts there. When
    # the boundary was exactly Shabbos the loop already took it and this comes out empty.
    gap_start = d - 7 + 1
    gap_end = end_serial - 1
    if gap_start <= gap_end and _any_regular_day(gap_start, gap_end, s, tables.special_days):
        parsha = has_parsha(end_serial, s, tables) or has_yom_tov(end_serial, s, tables.special_days)
        weeks.append(Week(end_serial, date_from_serial(end_serial), parsha, has_special_parsha(end_serial, s)))

    # d is now the first Saturday after the boundary, which is exactly the date the
    # incoming season's own loop would land on as its first candidate. A Chol Hamoed
    # Shabbos there becomes a row here instead of there. It is folded into the trailing gap
    # row above rather than added separately whenever that row already carries the same
    # holiday's name: the two stretches are the run-up to, and the middle of, one Yom Tov,
    # so the chart carries one row for it rather than a pair of identical ones.
    chol_hamoed_name = _holiday_name_for(d, s, tables.special_days) if _is_chol_hamoed_anchor(d, s, tables.special_days) else ""
    if chol_hamoed_name and (not weeks or weeks[-1].parsha != chol_hamoed_name):
        weeks.append(Week(d, date_from_serial(d), chol_hamoed_name, ""))

    return SeasonWeeks(start_serial, end_serial, weeks)


def split_choref_at_spring_cutover(weeks, s) -> int:
    """Where a שבת חורף season's weeks cross the spring DST cutover, the second Sunday of
    March, not the autumn one near the season's start. From that week on the season needs
    an actual שבת קיץ chart: the shul still davens on a summer schedule through Pesach once
    the clock springs forward. Returns len(weeks) if the whole season is before it, which
    should not happen in practice, Pesach always falling after the second Sunday of March.
    """
    for i, w in enumerate(weeks):
        if in_spring_dst_window(w.date, s):
            return i
    return len(weeks)


def next_available_year_for(season: str, s, today_serial: int | None = None) -> int:
    """The smallest Hebrew year for this season whose range has not fully elapsed, so the
    Generate form's year field never defaults to a season already over."""
    today = excel_serial(date.today()) if today_serial is None else today_serial
    y = hebrew_date_extended(today, s.use_gregorian_before_1582).year - 1
    for _ in range(6):
        if season_end_serial(season, y) >= today:
            break
        y += 1
    return y


def default_season_and_year(s, today_serial: int | None = None):
    """Which season and year Generate should open on: the next season chronologically
    after whichever contains today. A schedule is always prepared ahead of time for the
    coming season, never the one in progress and never one already over."""
    today = excel_serial(date.today()) if today_serial is None else today_serial
    y0 = hebrew_date_extended(today, s.use_gregorian_before_1582).year
    sukkos_y0 = date_from_hebrew(15, 7, y0)
    pesach_y0 = date_from_hebrew(15, 1, y0)
    sukkos_y0_plus_1 = date_from_hebrew(15, 7, y0 + 1)

    if today < sukkos_y0:
        current_season, current_year = "kayitz", y0 - 1  # still in last cycle's קיץ
    elif today < pesach_y0:
        current_season, current_year = "choref", y0
    elif today < sukkos_y0_plus_1:
        current_season, current_year = "kayitz", y0
    else:
        current_season, current_year = "choref", y0 + 1

    if current_season == "choref":
        return "kayitz", current_year
    return "choref", current_year + 1
